Allow the usage of a string or int as a value in portMappings

cdrage opened the pull request:
This allows the use of specifying either (for example) "80" or 80 within
portMappings of the Kedge file.

kadel commented:
We should also have unit tests covering this.

cdrage commented:
@kadel Yup. I'll add them the next few days. Just pushing this out first as proof of conversion (and for integration tests to successfully pass)

kadel commented:
Code looks good.
This might be good inspiration for tests https://github.com/kubernetes/kubernetes/blob/dd819b5013c99f930a6d528e508a900fa7679c4f/staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go#L45

cdrage commented:
@kadel Thanks for the pointer! Tests have been added :+1:

cdrage commented:
@surajssd @kadel please review tests :100:  and then this will unblock https://github.com/kedgeproject/kedge/pull/435 so we can push that in.

surajssd commented on pkg/spec/deployment.go:
```diff
@@ -33,7 +33,7 @@ import (
 	err := yaml.Unmarshal(data, &deployment)
 	if err != nil {
-		return errors.Wrap(err, "could not unmarshal into internal struct")
+		return errors.Wrap(err, "DeploymentSpecMod could not unmarshal into internal struct")
```
For the end user it is just `Deployment`!

surajssd commented on pkg/spec/types.go:
```diff
@@ -61,7 +62,8 @@ type ServiceSpecMod struct {
 	// targetPort and protocol in the format '<port>:<targetPort>/<protocol>'
 	// +optional
-	PortMappings []string `json:"portMappings,omitempty"`
+	// PortMappings []string `json:"portMappings,omitempty"`
```
remove the comment!

surajssd commented on pkg/yaml/types_test.go:
```diff
@@ -0,0 +1,48 @@
+		input  string
+		result SliceArrayorIntArray
+	}{
+		{"{\"portMappings\":[3306]}", SliceArrayorIntArray([]string{"3306"})},
```
JIC: if you don't want to escape the double quotes, then you can use the back ticks to enclose the entire string like following:

https://github.com/kedgeproject/kedge/blob/b7154f7765c89d0f29c3881e70ccc187115ad4a9/pkg/spec/controller_test.go#L155-L162

surajssd commented on pkg/yaml/types_test.go:
```diff
@@ -0,0 +1,48 @@
+	}{
+		{"{\"portMappings\":[3306]}", SliceArrayorIntArray([]string{"3306"})},
+		{"{\"portMappings\":[\"3306\"]}", SliceArrayorIntArray([]string{"3306"})},
+		{"{\"portMappings\":[\"80:80\", \"443:443\"]}", SliceArrayorIntArray([]string{"80:80", "443:443"})},
```
Please add a test that has mix of string and int. something like this

```
portMappings: [80, "3306:3306"]
```

surajssd commented:
Do we need to create a type of array type? Like it is done in this PR at https://github.com/cdrage/kedge/blob/859d75eb80269f7455d925d3877b6ff76bfc22f2/pkg/yaml/types.go#L25-L26

Can't we reuse the type defined in `"k8s.io/apimachinery/pkg/util/intstr"` which we already import in [pkg/spec/resources.go](https://github.com/kedgeproject/kedge/blob/b7154f7765c89d0f29c3881e70ccc187115ad4a9/pkg/spec/resources.go#L34).

https://github.com/kedgeproject/kedge/blob/b7154f7765c89d0f29c3881e70ccc187115ad4a9/vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go#L43-L47

cdrage commented:
@surajssd Unfortunately it has to be an array as @containscafeine 's implementation of portMappings iterates over an array. Simplest and easiest solution. 

I'll go ahead and make updates to this PR.

cdrage commented:
To be honest, I found this solution easier than iterating through Kubernetes code. I'll add more tests too :+1:

cdrage commented:
@surajssd 

Gone ahead and updated the code. I now extract to an interface{} and then iterate through the data to make sure it's either an int or string. Check out `yaml/types.go`

Also added another test (for a mix of string + int).

Thanks again for going through and doing a thorough test!

containscafeine commented on pkg/yaml/types.go:
```diff
@@ -0,0 +1,58 @@
+/*
```
A bit confused, why do we need to have a separate directory for this? Why can we not make this a part of `pkg/spec/util.go`? Or, if we think this is more global that `pkg/spec`, then maybe create a root level `util` directory and put this there?

surajssd replied:
because as we encounter more yaml types we can keep adding them here, I think what is done here makes sense!

containscafeine replied:
Sure, then maybe a `types` directory with our custom types

cdrage replied:
Because it mainly uses yaml/json and it follows similar practices used in libcompose (although maybe we shouldn't use that as an example...)

containscafeine commented on pkg/yaml/types.go:
```diff
@@ -0,0 +1,58 @@
+
+	// We create an array of strings by extracting to an interface and then iterating and figuring
+	// out each type. As per: https://golang.org/pkg/encoding/json/#Unmarshal
+	// we have a "case" for "string" and "float64"
```
why do we have a float64 here instead int?

containscafeine replied:
Due to this, I think this is considered as valid input -
```yaml
services:
- portMappings:
  - 3306.42
```
and gets converted to -
```yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: database
  name: database
spec:
  ports:
  - name: database-3306
    port: 3306
    protocol: TCP
    targetPort: 3306
  selector:
    app: database
  type: NodePort
status:
  loadBalancer: {}
```
which is not the correct behavior, we should be throwing an error in this case :(

kadel replied:
agreed, and name says Int, float is not int ;-)

cdrage replied:
See the above comment :+1: 

Yes, I understand that it says float, that is the default behaviour for JSON unmarshal: https://golang.org/pkg/encoding/json/#Unmarshal

kadel replied:
This will lead to a lot of errors. It has to be string or int, float can't be a valid value.

kadel replied:
If filtering out the float, can't be handled here, then it has to be handled somewhere else.
But doing it in the unamshaller is the right place

kadel replied:
I've just tried it in Kubernetes, and IntOrString from apimachinery is currently not allowing float : `for: "asdf": json: cannot unmarshal number 3.3 into Go value of type int32`

containscafeine commented on pkg/yaml/types.go:
```diff
@@ -0,0 +1,58 @@
+			case string:
+				// string, for JSON strings
+				stringArray = append(stringArray, v)
+			}
```
how about a `default` case where we return an error that some random type was passed which we don't support?

cdrage replied:
Sure, that works. Only other type would be bool or something like nil... :+1: But yeah.

kadel commented on pkg/yaml/types.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// SliceArrayorIntArray represents a slice or an integer
+type SliceArrayorIntArray []string
```
Could we have 

```go
type StringOrInt []string
type SliceArrayorIntArray []StringOrInt

func (s *StringOrInt) UnmarshalJSON(data []byte) error {
....
```

or something similar if possible,
so we can use it in #454 for `appversion` key that should have single value of IntOrString

cdrage replied:
Do you mind if we add this when we implement #454 instead since it's unrelated to this current commit? Reason being is that each "key" needs it's own Unmarshaler and thus we would have to add a custom UnmarshalJSON for each "key" used in `types.go`, for example:

```golang

func (s *StringOrInt) UnmarshalJSON(data []byte) error {
....


func (s *SliceArrayorIntArray) UnmarshalJSON(data []byte) error {
....

etc.
```

Or I could open a new PR for this after this PR #455 is implemented for the new type.

kadel replied:
Another thing is that this it should be of type `int` instead of `string`. As the whole point of this is to parse string as an integer, so resulting value should be stored in `integer`

cdrage replied:
@kadel Actually.. the whole thing is to parse back into a string. Hence why the result is []string

Reason being is that our PortMappings implemention by @containscafeine checks for "80:80/tcp" for example within the string. Hence the reason why all int's are converting to string and not the other-way around.

The end result would be something like this: []string{"80", "80:80/tcp", "3306:3306"} etc.

Having the output as a []int64 or []int wouldn't be possible..

cdrage commented:
@surajssd @kadel See: https://golang.org/pkg/encoding/json/#Unmarshal

When you use unmarshal, even though it's just an int, it will unmarshal to float64, hence the float and removing all decimal places.

I'll go ahead and implement the other things, but ignore the "float" section, it's the only way to get it from Unmarshal. :+1:

cdrage commented:
Ready for another round of reviews @kadel @surajssd

kadel commented:
@cdrage I think we should use `k8s.io/apimachinery/pkg/util/intstr` it solve all our problems, and it is simple. 
Check my this on how to use it: https://github.com/kadel/kedge/commit/657594549ce14a0eb95dcf04239031e03292a2dc

cdrage commented:
Thanks for the link @kadel :) I'll try it out!

cdrage commented:
@kadel Thanks for the link. I've updated the code accordingly... it's *way* better than before.

Mind doing a quick-check and then I'll add some integration / unit tests?
